### DISCUSSION TITLE
Global exception handling

### DISCUSSION THREAD

**Discussion by AmSmart** [dotnet/maui#653] [Ideas]
With the new host model in .NET Maui, will we get a straightforward way of catching exceptions?

**Comment by Flaflo** [comment]
I have the same question and I cannot find any solution.

**Comment by hartez** [comment]
> With the new host model in .NET Maui, will we get a straightforward way of catching exceptions?

What sort of thing are you looking for? How would you like it to behave when exceptions occur?

**Reply by AmSmart** [reply]
@hartez, In other .NET applications, especially ASP.NET Core, we can set up a catch-all such that expectations that weren't anticipated can be handled at a general level to keep applications from crashing unexpectedly. This is the behaviour being described.

**Reply by Strypper** [reply]
> @hartez, In other .NET applications, especially ASP.NET Core, we can set up a catch-all such that expectations that weren't anticipated can be handled at a general level to keep applications from crashing unexpectedly. This is the behaviour being described.
> 
> 

I think he left the chat 🤣🤣

**Comment by ThomasArdal** [comment]
Storing errors to a remote logging service would be a common use-case already commonly used on Xamarin and similar products.

**Comment by TanayParikh** [comment]
Cross linking https://github.com/dotnet/aspnetcore/pull/40765 which handles the case of WPF/WinForms in Blazor Hybrid. If there's a canonical way to do global exception handling (like we have in WinForms/WPF), we need to ensure .NET MAUI Blazor Hybrid is able to integrate with it directly. 

WPF/WinForms:

```csharp
AppDomain.CurrentDomain.UnhandledException += (sender, error) =>
{
    MessageBox.Show(text: error.ExceptionObject.ToString(), caption: "Error");
};
```

cc/ @hartez

**Reply by mattleibow** [reply]
Technically we could do something like this.. Right now we just log the exception and then carry on. In most cases, the app will throw in the flow - for example, updating the UI from a background thread. But there are cases - like image loading that we can only really log the exception because the operation failed in a place that cannot be caught.

This will make it nice to see stack traces and things.

**Reply by janseris** [reply]
@mattleibow any progress on global exception handler?

**Reply by MJSanfelippo** [reply]
@mattleibow  any update?

**Reply by viniciusverasdossantos** [reply]
:(

**Reply by harrykontopoulos** [reply]
It would be nice to have the equivalent of this: https://learn.microsoft.com/en-us/dotnet/api/system.windows.application.dispatcherunhandledexception?view=windowsdesktop-7.0

For common exceptions like a HttpRequestException it feels redundant to add try and catch all over the code. Similarly, on Asp.Net we don't kill the process every time a request results in an unhandled exception. Also, it would prevent shutting down the UI thread unnecessarily and just let the user know.

**Reply by egvijayanand** [reply]
Tried this, but no response.

**Comment by SmartmanApps** [comment]
Took me a long time to find this out, but this (or your own variation) will catch all exceptions (including the ones thrown by Xamarin which normally go un-noticed ;-) ). I haven't tried this in MAUI yet,, but I presume it'll work there too.

```
public App()
{
AppDomain.CurrentDomain.FirstChanceException+=CurrentDomain_FirstChanceException;
...
}

private void CurrentDomain_FirstChanceException(object sender, System.Runtime.ExceptionServices.FirstChanceExceptionEventArgs e)
{
System.Diagnostics.Debug.WriteLine($"********************************** UNHANDLED EXCEPTION! Details: {e.Exception.ToString()}");
}

```

**Reply by nogginbox** [reply]
Works for me and catching some XAML exceptions that were previously being hidden and taking me ages to work out.

**Reply by ThomasArdal** [reply]
Will the `FirstChanceException` event catch any exceptions not logged through Microsoft.Extensions.Logging?

**Reply by nogginbox** [reply]
I've not tested it thoroughly, but it is catching lots of exceptions including ones that are not causing obvious app issues.

Here is the output from starting my app up. The exception messages logged by this method are preceded with 'Before Handled':

```
Exception thrown: 'System.ArgumentException' in WinRT.Runtime.dll
Before Handled - Error: System.ArgumentException: Source object type is not a projected type and does not inherit from a projected type. (Parameter 'value')
   at WinRT.CastExtensions.GetRefForObject(Object value)

Exception thrown: 'System.DllNotFoundException' in System.Private.CoreLib.dll
Before Handled - Error: System.DllNotFoundException: Unable to load DLL 'Microsoft.Graphics.Canvas.Text.dll' or one of its dependencies: The specified module could not be found. (0x8007007E)
   at System.Runtime.InteropServices.NativeLibrary.LoadByName(String libraryName, QCallAssembly callingAssembly, Boolean hasDllImportSearchPathFlag, UInt32 dllImportSearchPathFlag, Boolean throwOnError)

Exception thrown: 'System.Runtime.InteropServices.COMException' in System.Private.CoreLib.dll
Before Handled - Error: System.Runtime.InteropServices.COMException (0x80070000): The operation completed successfully. (0x80070000)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)

Exception thrown: 'System.IO.FileNotFoundException' in System.Private.CoreLib.dll
Before Handled - Error: System.IO.FileNotFoundException: Element not found.
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHR(Int32 errorCode)
```

**Reply by TanayParikh** [reply]
> Will the `FirstChanceException` event catch any exceptions not logged through Microsoft.Extensions.Logging?

From my (limited) testing, it appears to be the case. 

Issue with this approach is it's a "firehose" of exceptions, as @NogginBox mentioned "it is catching lots of exceptions including ones that are not causing obvious app issues". This may work as an interim solution, but ideally we can integrate with `AppDomain.CurrentDomain.UnhandledException` to ensure we only catch critical exceptions.

Per the docs, the `FirstChanceException` "Occurs when an exception is thrown in managed code, before the runtime searches the call stack for an exception handler in the application domain.". [Source](https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.firstchanceexception?view=net-6.0).

**Reply by ThomasArdal** [reply]
It sounds a bit crazy to me that Maui supports both Microsoft.Extensions.Logging and events like `FirstChanceException` and `UnhandledException` and then don't log all exceptions through Microsoft.Extensions.Logging. If Microsoft.Extensions.Logging is supported in the first place, the events should not be intended for logging but other features like showing error messages to the user or similar. If we still need to subscribe to both events and logging to log all errors, the clever thing about having a logging framework in the mix goes away (IMO).

I think it's fine that Maui logs exceptions when they happen. But if they are not routed through the logging framework, it's way harder to filter out un-interesting exceptions like we can do with namespace filtering and similar in Microsoft.Extensions.Logging.

**Reply by JayMansel** [reply]
Just tested within MAUI and works beautifully :) Should definitely be the selected answer (go on OP :P)

**Reply by mattjohnsonpint** [reply]
You probably don't want to log the firehose that is first chance exceptions.   First chance means immediately after throwing,  even if it's caught and handled.

Take a look here for a better solution.

https://github.com/dotnet/maui/discussions/653#discussioncomment-3434251

**Comment by mattjohnsonpint** [comment]
In putting together [Sentry error monitoring for .NET MAUI](https://sentry.io/for/dotnet-maui/), we've faced this head-on.

There are several different issues to deal with, as each platform that MAUI supports has its own quirks.

Here's a class you can use in your own code that provides a global event handler that works on all major platforms. (I didn't test Tizen.)
https://gist.github.com/mattjohnsonpint/7b385b7a2da7059c4a16562bc5ddb3b7

Usage:

```csharp
MauiExceptions.UnhandledException += (sender, args) => { /* your code */ }
```

(If you decide to use Sentry, this is built-in.)

**Reply by MauriceChocoSwiss** [reply]
I've tried to use it, but I didn't find where to place it. I've placed it in the CreateMauiApp function but it didn't trigger when I throw Exception from iOS. 

Can you tell me where I need to use it ?

**Reply by headintheclouds21** [reply]
I've also tried this for MacCatalyst and was unable to trigger it when I'm throwing an Exception. 

Could you advise on where to use it in the application?

**Reply by anpin** [reply]
we are porting our app from XF to MAUI and we are using Sentry. Original app had catchers implemented in the same way as the gist. However MAUI version on iOS crashes in ```ApplyRelativeSourceBininding``` without catching anything or logging it to Sentry.

**Comment by MauriceChocoSwiss** [comment]
Any news ? 
I've tested the solutions I've found here but it's not exactly what I want to do. 

I would like to do like in Flutter (https://docs.flutter.dev/testing/errors) and catch all kind of exceptions (even customs I've made) and display them in Error page but the solutions I've found here doesn't help me to do that.

**Reply by ooleglysiak** [reply]
It works for me:
https://gist.github.com/mattjohnsonpint/7b385b7a2da7059c4a16562bc5ddb3b7

**Reply by MauriceChocoSwiss** [reply]
Thank, I've just found and test this way of catch error this week, I've not tested yet for iOs. 
I've just trouble with Android in debug, with the Exception handler, he still close my app, but only in debug, in release it work like a charm !

Anyway thank you !!

**Reply by 7702244** [reply]
> It works for me:
> https://gist.github.com/mattjohnsonpint/7b385b7a2da7059c4a16562bc5ddb3b7

This is good for logging, but will not stop the application from crashing.

**Reply by ooleglysiak** [reply]
> > It works for me:
> > https://gist.github.com/mattjohnsonpint/7b385b7a2da7059c4a16562bc5ddb3b7
> 
> This is good for logging, but will not stop the application from crashing.

You can use the same methods as in Xamarin(Example:https://peterno.wordpress.com/2015/04/15/unhandled-exception-handling-in-ios-and-android-with-xamarin/)

**Reply by MauriceChocoSwiss** [reply]
I do these in each project main CS file and I've access to Handled property to set to true and prevent the app from crashing.

@ooleglysiak , with the link you shared, the guy only explain how to catch and log but not how to prevent app from crashing.

**Reply by ooleglysiak** [reply]
Sorry, sent the wrong link, didn't notice

**Reply by MauriceChocoSwiss** [reply]
I didn't get it work with iOS... I can't prevent app from crashing instead of showing a little snack to tell me something append... It's insane to crash the app when a api call return an exception (didn't find a way to catch all my api calls without puting try/catch arround each) or for example, when I lost internet connection and custom exception I want to throw

**Comment by viniciusverasdossantos** [comment]
Has anyone got a solution to prevent the application from closing when an exception occurs?

**Reply by MauriceChocoSwiss** [reply]
Yop, I've found this for windows :
```
        Current.UnhandledException += (s, a) =>
        {
            a.Handled = true;
        };   
```  
 => to be placed in the App in Platforme => Windows => App.xaml.cs

And for Android :
```
    public MainApplication(IntPtr handle, JniHandleOwnership ownership)
        : base(handle, ownership)
    {
        AndroidEnvironment.UnhandledExceptionRaiser += AndroidEnvironment_UnhandledExceptionRaiser;
    }
    
    protected override MauiApp CreateMauiApp() => MauiProgram.CreateMauiApp();
    
    private static void AndroidEnvironment_UnhandledExceptionRaiser(object sender, RaiseThrowableEventArgs e)
    {
        e.Handled = true;
    }
```
 => In Plateform => Android =>MainApplication.cs => Work only in release, in debug it crash app 
 
 For iOS, I didn't found a way to achieve this...

**Reply by MauriceChocoSwiss** [reply]
.Net 7, you need to add it to the App method, before the initialize

**Reply by headintheclouds21** [reply]
Any idea how to do this on MacCatalyst with .NET 7?

**Reply by MauriceChocoSwiss** [reply]
My project doesn't include MacCatalyst. You need to look at MacOS documentation to know if it's like on iOS (if unhandled exception is throwned, crash the app)

**Reply by Strypper** [reply]
Thanks you I but I my poup service in sitting at MAUI application, can't invoke the service from Platform specific

**Reply by janseris** [reply]
> Thanks you I but I my poup service in sitting at MAUI application, can't invoke the service from Platform specific

Create a service through which you will Invoke the popup and inject it into the platform code

**Comment by SmartmanApps** [comment]
Thanks @TanayParikh  and @mattjohnsonpint, I didn't know about UnhandledException - that'll teach me to read through ALL the Intellisense options when I'm trying something new! :-)

**Comment by MJSanfelippo** [comment]
Any updates on this? None of the given solutions work incredibly well, or only work in some scenarios/platforms. This seems like a super important feature to have for such a high-profile platform, but there seems to be no straight-forward way of doing it.

**Reply by Xyncgas** [reply]
how do you catch stackoverflow

**Reply by Xyncgas** [reply]
Although, if there's some sort of 'main' entry to your application, can't you just put try catch around it

**Comment by viniciusverasdossantos** [comment]
In maui .net 8 was there any solution? :(

**Reply by Strypper** [reply]
LOL

**Comment by rf-0** [comment]
Also interested.

**Reply by plppp2001** [reply]
Yeah me too...

**Reply by Dumuzy** [reply]
Anything new about the global exception handler?  Such a thing is clearly needed. The XAML interpreter e.g. crashes under some circumstances in release mode but not in debug with the exactly same code. How could we debug this stuff? Best would be a catch all exception handler with which we could log the exception.

**Comment by CfmHansVallee** [comment]
Any update on this?
I am using Maui with net8, the code works fine, but the app still crashes, really looking forward to a true global exception handling pattern.

**Comment by VagnerVit** [comment]
I am trying to catch ObjectDisposedException which causes the whole app to crash and I think it causes the image disposing... I really cannot modify the base MAUI Image component.

Any updates?

**Comment by developer9969** [comment]
Hi just to say that this would be a dream in Maui land but sorry to be cynical but will never happen.

**Comment by pme442** [comment]
Any updates on this yet??

**Reply by janseris** [reply]
They don't react to discussions

**Comment by FreakyAli** [comment]
For an app targetting iOS and Android this should be more than enough: 


```C#
public static class GlobalExceptionHandler
{
    // We'll route all unhandled exceptions through this one event.
    public static event UnhandledExceptionEventHandler UnhandledException;

    static GlobalExceptionHandler()
    {
        AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
        {
            UnhandledException?.Invoke(sender, args);
        };

        // Events fired by the TaskScheduler. That is calls like Task.Run(...)     
        TaskScheduler.UnobservedTaskException += (sender, args) =>
        {
            UnhandledException?.Invoke(sender, new UnhandledExceptionEventArgs(args.Exception, false));
        };

#if IOS

        // For iOS and Mac Catalyst
        // Exceptions will flow through AppDomain.CurrentDomain.UnhandledException,
        // but we need to set UnwindNativeCode to get it to work correctly. 
        // 
        // See: https://github.com/xamarin/xamarin-macios/issues/15252
        
        ObjCRuntime.Runtime.MarshalManagedException += (_, args) =>
        {
            args.ExceptionMode = ObjCRuntime.MarshalManagedExceptionMode.UnwindNativeCode;
        };

#elif ANDROID

        // For Android:
        // All exceptions will flow through Android.Runtime.AndroidEnvironment.UnhandledExceptionRaiser,
        // and NOT through AppDomain.CurrentDomain.UnhandledException

        Android.Runtime.AndroidEnvironment.UnhandledExceptionRaiser += (sender, args) =>
        {
            args.Handled = true;
            UnhandledException?.Invoke(sender, new UnhandledExceptionEventArgs(args.Exception, true));
        };

        Java.Lang.Thread.DefaultUncaughtExceptionHandler = new CustomUncaughtExceptionHandler(e =>
            UnhandledException?.Invoke(null, new UnhandledExceptionEventArgs(e, true)));
#endif
    }
}

#if ANDROID
public class CustomUncaughtExceptionHandler(Action<Java.Lang.Throwable> callback)
    : Java.Lang.Object, Java.Lang.Thread.IUncaughtExceptionHandler
{
    public void UncaughtException(Java.Lang.Thread t, Java.Lang.Throwable e)
    {
        callback(e);
    }
}
#endif
```

And a simple way to use this would be: 

```
GlobalExceptionHandler.UnhandledException += GlobalExceptionHandler_UnhandledException;

 private void GlobalExceptionHandler_UnhandledException(object sender, UnhandledExceptionEventArgs e)
  {
      var exception = e.ExceptionObject as System.Exception;
      // Log this exception or whatever
  }
```

**Reply by ThomasArdal** [reply]
Isn't this the same as [MauiExceptions](https://gist.github.com/mattjohnsonpint/7b385b7a2da7059c4a16562bc5ddb3b7) or did you optimize something?

**Reply by FreakyAli** [reply]
It has `TaskScheduler.UnobservedTaskException` and native Android uncaught thread exceptions so I guess it gets you a bit more

**Comment by stephenquan** [comment]
I’m removing my answer because it shouldn’t be used in production. The code is not safe under concurrent, multithreaded access to the Preferences object and can cause crashes, and it also rarely catches exceptions in Android and iOS release builds.

<strike>
You can persist unhandled exceptions in [Preferences](https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/storage/preferences?view=net-maui-9.0&tabs=windows) with:

```c#
// Install a global exception handler to persist the crash information before the app crashes.
#if WINDOWS
global::Microsoft.Maui.MauiWinUIApplication.Current.UnhandledException += (sender, e) =>
{
    if (e.Exception is Exception ex)
    {
        Preferences.Default.Set("LastCrashTime", DateTimeOffset.Now.ToUnixTimeMilliseconds());
        Preferences.Default.Set("LastCrashMessage", ex.Message);
        Preferences.Default.Set("LastCrashStackTrace", ex.StackTrace);
    }
    e.Handled = false;
};
#else
global::System.AppDomain.CurrentDomain.UnhandledException += (sender, e) =>
{
    if (e.ExceptionObject is Exception ex)
    {
        Preferences.Default.Set("LastCrashTime", DateTimeOffset.Now.ToUnixTimeMilliseconds());
        Preferences.Default.Set("LastCrashMessage", ex.Message);
        Preferences.Default.Set("LastCrashStackTrace", ex.StackTrace);
    }
};
#endif
```

Then, on the next run of the app, you can casually retrieve any exceptions from the last 24 hours, and choose what to do with it.

```c#
if (Preferences.Default.Get("LastCrashTime", 0L) is long lastCrashTime
    && DateTimeOffset.Now.ToUnixTimeMilliseconds() < lastCrashTime + 24 * 60 * 60 * 1000
    && Preferences.Default.Get("LastCrashMessage", string.Empty) is string lastCrashMessage
    && Preferences.Default.Get("LastCrashStackTrace", string.Empty) is string lastCrashStackTrace)
{
    Trace.WriteLine($"The app crashed on {DateTimeOffset.FromUnixTimeMilliseconds(lastCrashTime).LocalDateTime} with the following message:\n\n{lastCrashMessage}\n\n{lastCrashStackTrace}");
    // Shell.Current.GotoAsync(nameof(YourAppRecoveredFromASeriousCrashPage));
}
```

</strike>

**Comment by mwilson25** [comment]
This thread is probably getting a little old and stale but here goes.

The global unhandled exception handler provided by mattjohnsonpint seems to work as advertised.

I have tested by putting the code

throw new Exception("Test Exception")

on one of my button handlers.

My problem is the IOS version of my app (all OK on Windows). 

While the unhandled exception event does fire on my IOS app, it does so only after

UIApplication.Main(args, null, typeof(AppDelegate)) in the mainline has already exited with an exception.

What I found is that despite my best efforts (and I tried several ways)

1. I am not able to inform my users via an alert message that the app has gone sideways.
2. I am not able to write error details to my sqlite database

Using the despatcher did not work.

Now from the comments so far on this thread, this has been a problem for some time. Its now about time that we have the ability to intercept an unhandled exception (caused by a simple throw statement) and provide basic information to the screen to inform the user!

Something needs to be done to rectify this now.

Can anyone help me.

**Comment by stephenquan** [comment]
In mattjohnsonpint code, he flows the exceptions to AppDomain.CurrentDomain.UnhandledException, but have you tried reacting to the exception without flowing thru? e.g.

```c#
#if IOS || MACCATALYST
ObjCRuntime.Runtime.MarshalManagedException += (sender, args) => /* React to args.Exception */;
ObjCRuntime.Runtime.MarshalObjectiveCException += (sender, args) => /* React to args.Exception */;
#elif ANDROID
```

**Reply by mwilson25** [reply]
That's a good suggestion.

I had already tried with ObjCRuntime.Runtime.MarshalManagedException += (sender, args) => /* React to args.Exception */;

which unfortunately triggers after  

UIApplication.Main(args, null, typeof(AppDelegate)) in the mainline has exited with an exception. 

As a result I cannot seem to display an alert to my user.

I have just tried your second suggestion

The ObjCRuntime.Runtime.MarshalObjectiveCException += (sender, args) => /* React to args.Exception */;

which does not fire for me at all.

Perhaps that is because I am initiating the exception with a simple

throw new Exception ("Test Exception"); 

Not sure.

Anyway - its no go I am afraid.

**Reply by stephenquan** [reply]
If you're on iOS 26, you may like to try this. N.B. The code hasn't been fully validated. The code attempts, for iOS 26, to propogate managed exceptions as objective C exceptions.

```c#
#if IOS || MACCATALYST
ObjCRuntime.Runtime.MarshalManagedException += (_, args) =>
{
    if (System.OperatingSystem.IsIOSVersionAtLeast(26) || System.OperatingSystem.IsMacCatalystVersionAtLeast(26))
    {
        args.ExceptionMode = ObjCRuntime.MarshalManagedExceptionMode.ThrowObjectiveCException;
        return;
    }
    args.ExceptionMode = ObjCRuntime.MarshalManagedExceptionMode.UnwindNativeCode;
};
#endif
```
